password_enhancements.module

```diff
@@ -0,0 +1,403 @@
+  }
+
+  $form['#validate'][] = 'password_enhancements_form_user_register_form_alter_validate';
+}
+
+/**
+ * Validate callback for password_enhancements_form_user_register_form_alter().
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ *
+ * @throws \Drupal\Component\Plugin\Exception\InvalidPluginDefinitionException
+ * @throws \Drupal\Component\Plugin\Exception\PluginNotFoundException
+ * @throws \Drupal\Component\Plugin\Exception\PluginException
+ */
+function password_enhancements_form_user_register_form_alter_validate(array &$form, FormStateInterface $form_state): void {
+  if (!empty($password = $form_state->getValue('pass'))) {
+    $roles = $form_state->getValue('roles') ?? [];
+
+    /** @var \Drupal\password_enhancements\Entity\Storage\PolicyEntityStorageInterface $password_policy_storage */
+    $password_policy_storage = Drupal::entityTypeManager()->getStorage('password_enhancements_policy');
+    $policy = $password_policy_storage->loadByRoleAndPriority($roles);
+
+    $constraints = _password_enhancements_get_constraints($roles);
+    $error_messages = [];
+    $minimum_required_constraints = $policy->getMinimumRequiredConstraints();
+    $valid_count = 0;
+    foreach ($constraints as $constraint) {
+      /** @var \Drupal\password_enhancements\Plugin\PasswordConstraintPluginInterface $plugin_instance */
+      $plugin_instance = Drupal::service('plugin.manager.password_constraint')->createInstance($constraint->getType(), $constraint->getConfiguration());
+
+      if (!$plugin_instance->validate($password)) {
+        $error_messages[] = $plugin_instance->getErrorMessage();
+      }
+      else {
+        $valid_count++;
+      }
+
+      if ($minimum_required_constraints <= $valid_count) {
+        break;
+      }
+    }
+
+    // Display any validation error messages.
+    if ($error_messages !== [] && $valid_count < $minimum_required_constraints) {
+      $messages = [
+        '#theme' => 'password_enhancements_error_messages',
+        '#error_message' => t("The given password doesn't meet some requirements:"),
+        '#messages' => $error_messages,
+      ];
+      $form_state->setError($form['account']['pass'], Drupal::service('renderer')->render($messages));
+    }
+  }
+}
+
+/**
+ * Submit callback for password_enhancements_form_user_register_form_alter().
+ *
+ * This callback runs only if the password field was set to always require and
+ * the verify_email set to true.
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ *
+ * @throws \Drupal\Core\Entity\EntityStorageException
+ *   If the user was failed to save.
+ */
+function password_enhancements_form_user_register_form_alter_password_submit(array &$form, FormStateInterface $form_state): void {
+  /** @var \Drupal\user\UserInterface $user */
+  $user = $form_state->getFormObject()->getEntity();
```

Comment:
> This callback runs only if the password field was set to always require and
> * the verify_email set to true.

1. The name of the function does not indicate this.
2. If it should only run in this given circumtenses I would rather check in this function whether it should do anything or not instead trusting a developer that it does not call when the given statement is not true. IOW, let's move the check from `password_enhancements_form_user_register_form_alter()` to here and always add this submit handler to the form.

Reply:
I didn't wanted to duplicate the condition, it is currently called from only one place.

I've renamed the function to `password_enhancements_form_user_register_form_alter_password_required_submit()`

Reply:
I did not suggest duplication, I suggested to check in this function whether there is anything to do and remove the check from the form alter, IOW always add this submit handler, so other alter hooks could see that it gets called.

Reply:
I also wanted to avoid calling it unless it's necessary, but I guess it makes sense, I'm going to change it.

Reply:
Actually I can't add the submit/validate callback always, because if the password field is missing (iow not necessary for the registration) it is not set, I've only added a condition to the submit callback to check whether the password field exist or not.